### ISSUE TITLE
`IndexType` not implemented for `u64`

### ISSUE THREAD

**Issue by sanmai-NL** [bluss/petgraph#190]
[`IndexType`](https://docs.rs/petgraph/0.4.10/petgraph/graph/trait.IndexType.html)

Of course a graph of size in excess of the maximum representable `u32` would be extremely large and questionable, but the trait is implemented for `usize` already.

**Comment by bluss** [comment]
Right, the idea is that an index is always representable in `usize`. Now that has not quite come true (16-bit platforms of Rust). The current design requires lossless conversion between IndexType <=> usize. It also says that you should use `usize` if you need a bigger graph than `u32`.

**Comment by bluss** [comment]
Since it is by design, closing.

**Comment by sanmai-NL** [comment]
@bluss: okay, what’s the impact of using a `usize` node index type in a context where graphs are deserialized across machines with different bitness?

**Comment by bluss** [comment]
The IndexType is known when serializing a graph, but also when deserializing a graph. In general in serde, we require the data must be deserialized using the same type as it was serialized with. petgraph has some de facto loosening of those restrictions, but it also depends on the serializer used.

The Node/Edge indices are de/serialized like plain unsigned integers that match the IndexType. Then the graph node, edge counts is checked so that it can be represented with the given IndexType. If there's a mismatch the serialization returns an error.

With a serializer like JSON, we can deserialize as any index type as long as the chosen type can hold all the indices of the graph, because all integers serialize identically.

With bincode, it always encodes usize like u64. Bincode being a binary format also depends on the exact widths of integer types, so the graph must be deserialized using exactly the IndexType that it was serialized with. It is my understanding that de/serializing using `usize` should be portable everywhere as well, as long as the number of nodes, edges in a graph fits inside the type limits where it is deserialized.

The current serialization code it deserializes some `Vec` of nodes or edges before checking their count vs type limits. I'm unsure how the serializers handle this, if the number of elements exceeds the platform usize or does not fit in memory -- I don't know whether they error nicely or blow up in some unpleasant way.

**Comment by bluss** [comment]
Serialization Compatibility that I know

- a Graph can always be deserialized as a StableGraph
- a StableGraph with no vacancies can be deserialized as a Graph (otherwise there is an error)
- For `GraphType<N, E, Ty, Ix>` where GraphType is `Graph` or `StableGraph`
  - `Ty` must be exactly the same (directed or undirected)
  - `N`, `E` can be replaced with anything that deserializes from the same representation (type and serializer dependent)
  - `Ix` can be replaced with anything that deserializes from the same representation (type and serializer dependent). The chosen index type must also be able to hold the node and edge counts (including vacancies for StableGraph) for the particular graph being deserialized.
     - As we noted before, JSON is flexible about integer widths and bincode is not.

**Comment by bluss** [comment]
Point in the code for serialization. The code uses separate ser/de representations for no-copy serialization and deserialization into exactly (mostly) the target representation. 

https://github.com/bluss/petgraph/blob/master/src/graph_impl/serialization.rs#L18

**Comment by bluss** [comment]
We should probably lift this serialization info into the docs
